### ISSUE TITLE
State is not cleared on connect

### ISSUE THREAD

**Issue by peppy** [ppy/osu-server-spectator#21]
```
User 7010695 ending play session (Beatmap:2643109 Mods: Ruleset:0) fail: Microsoft.AspNetCore.SignalR.Internal.DefaultHubDispatcher[8] Failed to invoke hub method 'JoinRoom'. MySqlConnector.MySqlException (0x80004005): Duplicate entry '38413-12918535' for key 'multiplayer_rooms_high.multiplayer_rooms_high_room_id_user_id_unique' ---> MySqlConnector.MySqlException (0x80004005): Duplicate entry '38413-12918535' for key 'multiplayer_rooms_high.multiplayer_rooms_high_room_id_user_id_unique' at MySqlConnector.Core.ServerSession.ReceiveReplyAsyncAwaited(ValueTask`1 task) in /_/src/MySqlConnector/Core/ServerSession.cs:line 817 at MySqlConnector.Core.ResultSet.ReadResultSetHeaderAsync(IOBehavior ioBehavior) in /_/src/MySqlConnector/Core/ResultSet.cs:line 49 at MySqlConnector.MySqlDataReader.ActivateResultSet(CancellationToken cancellationToken) in /_/src/MySqlConnector/MySqlDataReader.cs:line 118 at MySqlConnector.MySqlDataReader.CreateAsync(CommandListPosition commandListPosition, ICommandPayloadCreator payloadCreator, IDictionary`2 cachedProcedures, IMySqlCommand command, CommandBehavior behavior, IOBehavior ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/MySqlDataReader.cs:line 437 at MySqlConnector.Core.CommandExecutor.ExecuteReaderAsync(IReadOnlyList`1 commands, ICommandPayloadCreator payloadCreator, CommandBehavior behavior, IOBehavior ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/Core/CommandExecutor.cs:line 60 at MySqlConnector.MySqlCommand.ExecuteNonQueryAsync(IOBehavior ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/MySqlCommand.cs:line 264 at Dapper.SqlMapper.ExecuteImplAsync(IDbConnection cnn, CommandDefinition command, Object param) in /_/Dapper/SqlMapper.Async.cs:line 654 at osu.Server.Spectator.Hubs.MultiplayerHub.UpdateDatabaseParticipants(MultiplayerRoom room) in /Users/dean/Projects/osu-server-spectator/osu.Server.Spectator/Hubs/MultiplayerHub.cs:line 422 at osu.Server.Spectator.Hubs.MultiplayerHub.JoinRoom(Int64 roomId) in /Users/dean/Projects/osu-server-spectator/osu.Server.Spectator/Hubs/MultiplayerHub.cs:line 94 at lambda_method(Closure , Object ) at Microsoft.AspNetCore.SignalR.Internal.DefaultHubDispatcher`1.ExecuteHubMethod(ObjectMethodExecutor methodExecutor, THub hub, Object[] arguments) at Microsoft.AspNetCore.SignalR.Internal.DefaultHubDispatcher`1.<>c**DisplayClass13_0.<
```

Looking into this, it would seem that the case is where a user reconnects to the multiplayer hub before their previous connection times out (and triggers a disconnection server-side). This means the user state (which is stored against only user_id) may be carried across to the new user as it is only removed on disconnection.

It also means that the reverse can happen: a user can reconnect and start using the state, only to have their previous connection disconnect and clear it.

This is a pretty critical oversight.

**Comment by peppy** [comment]
This one goes pretty deep. I've started work on it (better locking state dictionary for a start, see https://github.com/ppy/osu-server-spectator/compare/master...peppy:improve-state-reliability?expand=1) but hit another issue in the process:

We want to clean up state of the previous connection when the same user connects a second time. This isn't possible because the whole `MultiplayerHub` performs operations on the *local user context*. This means that a call to `LeaveRoom` will run for the local user in the context of their current connection, not the previous one.

While in theory this will still work correctly (except for removing connections from SignalR rooms), I'm still a bit iffy about leaving it in this state. Not sure how it will scale going forward.

I've deployed (most of) the changes I made on that linked branch to the production server as I'm pretty sure it will be in a better state than it was, but going to take some more time to think this one through before PRing.

**Comment by bdach** [comment]
I'm not sure if I'm understanding the issue correctly from this description. My immediate idea for a solution seems to match what you have there (adding a connection ID to the user context), but not sure where the scalability worries are coming from - is it the fact that it has to be sent from the client every time?

**Comment by peppy** [comment]
I didn't mean scaling in terms of performance, but in terms of how we write methods. May have been a bit unclear.

**Comment by peppy** [comment]
Resolved via #25.
